### PR TITLE
Speed up deserialization of large vectors by a factor of ten

### PR DESCRIPTION
**Pull request by bgamari**
tl;dr. the current `Get` implementation is rather slow. This speeds it up
significantly but I thought it would be appropriate to give you the opportunity
to object.

Previously the Core produced by `generateM` was quite poor. So poor, in fact,
that it would allocate a cons cell for every element it would deserialize. Take,
for instance, the STG for the inner loop of the deserializer
```hs
let {
  $s$wa1_s90Z :: Int#
              -> [Int]
              -> State# s1_a61e
              -> (# State# s1_a61e, Int #) =
      sat-only \r srt:SRT:[] [sc_s910 sc1_s911 sc2_s912]
          case sc1_s911 of
          _
          { [] ->
                let {
                  sat_s914 :: Int = NO_CCS I#! [sc_s910];
                } in 
                  (#,#) [sc2_s912 sat_s914];

            : x2_s915 xs1_s916 ->
                case x2_s915 of _
                { I# x#1_s918 ->
                      case writeIntArray# [ipv1_s90Y sc_s910 x#1_s918 sc2_s912]
                      of s'#_s919
                      { __DEFAULT ->
                            case +# [sc_s910 1]
                            of sat_s91a
                            { __DEFAULT -> $s$wa1_s90Z sat_s91a xs1_s916 s'#_s919; };
                      };
                };
          };
} in 
  case
      $s$wa1_s90Z 0 a3_s90M ipv_s90X
  of ...
```
Yes, that is a cons list there. The C-- this produces is about as bad as you
would expect.

The new implementation appears to be free of any allocation and in the case of
unboxed vectors of `Int`s, performs significantly better,

| N   | Before        | After               |
| --- | ------------- | ------------------- |
| 3   | 368 ± 73ns    | 335 ± 60ns          |
| 30  | 1.5 ± 0.350μs | 1.3 ± 0.17μs        |
| 300 | 10.0 ± 1μs    | 8.7 us ± 1μs        |
| 3e4 | 6.1 ± 0.6ms   | 980μs ± 144μs       |
| 3e5 | 80 ± 4.2ms    | 8.2 ± 1ms           | 

Instead of trying to convince vector's stream fusion framework to produce
reasonable code here I simply rewrote the serializer as the imperative loop that
we want. While the code is clearly deeply unsafe, the Core that is produced is
correct on 7.10.3, 7.8.4, and 8.0.1. Moreover, it performs equivalently to the
old implementation on small inputs and nearly an order of magnitude faster on
large (e.g. 300000 elements) inputs.

### CODE REVIEW

**Comment by bgamari** [comment]
Alright, I'm going to go ahead and merge this.
